### PR TITLE
feat(change_detection): add mode to ChangeDetector

### PR DESCRIPTION
**Pull request by vsavkin**
Currently, we walk the whole tree of change detectors every single time. 
- We should be able to "detach" a change detector subtree temporary and then reattach it back. This is useful for such directives as [hidden] and [throttle].

```
class Hidden {
    cd:ChangeDetector;

    set hidden(val) {
        cd.mode = val ? CHECK_ALWAYS : DETACHED;
    }
}
```

The DETACHED change detector will be skipped.
- We should be able to take advantage of components that depend only on immutable bindings. So we want to check them once, when one the bindings changes. This is fairly straightforward. The component can change only when onChange is called, which happens during a change detection run.  

```
class OnlyImmutableBindings {
    cd:ChangeDetector;

    onChange(_) {
        cd.mode = CHECK_ONCE;
    }
}
```
- We should be able to take advantage of components that depend only observable objects. This is trickier. 

Consider the following example, where Comp3 depends only on observable bindings:

```
Comp1
  |
Comp2
  |
Comp3
```

When one of the observables Comp3 depends on fires an event, we should set Comp1, Comp2, and Comp3 to CHECK_ONCE. 

```
class OnlyObservableBindings {
    cd:ChangeDetector;

    onChange(changes) {
        changes.forEach((c) => {
            stopObserving(c.previousValue);
            observe(c.currentValue, cd.markAsCheckOnce);
        })
    }
}
```
- I do not think we should be able to inject ChangeDetector into a component. Some other object will eventually be provided. This PR does not add such an object.
- Such mixins as `OnlyObservableBindings` and `OnlyImmutableBindings` can easily be created. 

@yjbanov @mhevery can you guys take a look?

### CODE REVIEW

File: modules/change_detection/src/abstract_change_detector.js [hunk 1]
```diff
@@ -1,12 +1,15 @@
+import {isPresent} from 'facade/lang';
 import {List, ListWrapper} from 'facade/collection';
-import {ChangeDetector} from './interfaces';
+import {ChangeDetector, CHECK_ALWAYS, CHECK_ONCE, CHECKED, DETACHED} from './interfaces';
 
 export class AbstractChangeDetector extends ChangeDetector {
   children:List;
   parent:ChangeDetector;
+  status:string;
```
Comment (vicb) [inline modules/change_detection/src/abstract_change_detector.js]: mode ?

Comment (vsavkin) [inline modules/change_detection/src/abstract_change_detector.js]: Yes. I believe so. 

Since currently we do not use interfaces, we technically do not have to do it. But once we start using them, we will have to do it.

File: modules/change_detection/src/interfaces.js [hunk 2]
```diff
@@ -19,12 +19,38 @@ export class ChangeRecord {
   }
 }
 
+
+/**
+ * CHECK_ONCE means that after calling detectChanges the mode of the change detector
+ * will become CHECKED.
+ */
+export const CHECK_ONCE="CHECK_ONCE";
```
Comment (yjbanov) [inline modules/change_detection/src/interfaces.js]: I think we should use typed constants here so we could transpile them to enums with minimal breaking changes.

File: modules/change_detection/src/abstract_change_detector.js [hunk 3]
```diff
@@ -31,8 +34,20 @@ export class AbstractChangeDetector extends ChangeDetector {
   }
 
   _detectChanges(throwOnChange:boolean) {
+    if (this.mode === DETACHED || this.mode === CHECKED) return;
+
     this.detectChangesInRecords(throwOnChange);
     this._detectChangesInChildren(throwOnChange);
+
+    if (this.mode === CHECK_ONCE) this.mode = CHECKED;
```
Comment (yjbanov) [inline modules/change_detection/src/abstract_change_detector.js]: How do we want to handle situations when one of the above methods throw?

Comment (vsavkin) [inline modules/change_detection/src/abstract_change_detector.js]: I think we should do nothing. 

- If the mode is CHECK_ALWAYS, then it should remain CHECK_ALWAYS.
- If the mode is CHECK_ONCE, then it should remain CHECK_ONCE. We cannot change it to CHECKED because not all bindings have been checked.

File: modules/change_detection/src/interfaces.js (hunk 2 above)
Comment (vsavkin) [inline modules/change_detection/src/interfaces.js]: Let's talk about it tomorrow. This is not the only place where we define constants in this way. So if we decide to change it, we should probably do it everywhere.

File: modules/change_detection/src/abstract_change_detector.js [hunk 4]
```diff
@@ -31,8 +34,20 @@ export class AbstractChangeDetector extends ChangeDetector {
...
+  }
+
+  markPathToRootAsCheckOnce() {
```
Comment (rkirov) [inline modules/change_detection/src/abstract_change_detector.js]: Feels more like a static method than a member method.

File: modules/change_detection/src/abstract_change_detector.js [hunk 5]
```diff
@@ -31,8 +34,20 @@ export class AbstractChangeDetector extends ChangeDetector {
...
+    var c = this;
+    while(isPresent(c) && (c.mode === CHECKED || c.mode === CHECK_ALWAYS)) {
```
Comment (rkirov) [inline modules/change_detection/src/abstract_change_detector.js]: Don't you want `c.mode !== DETACHED` instead of `(c.mode === CHECKED || c.mode === CHECK_ALWAYS))` to allow for `CHECK_ONCE`. Remember if we give an imperative API, someone might put an intermediate CD into `CHECK_ONCE` without correspondingly putting all parents into `CHECK_ONCE`.

File: modules/change_detection/src/interfaces.js (hunk 2 above)
Comment (rkirov) [inline modules/change_detection/src/interfaces.js]: missing spaces '=' -> ' = '

**Comment by mhevery** [comment]
LGTM,

Do we have a better name for `ChangeDetector`? How about `Binder`, `Bindings`, ????

File: modules/change_detection/test/change_detection_spec.js [hunk 6]
```diff
@@ -466,6 +467,75 @@ export function main() {
...
+            var grandChild = changeDetector(CHECKED, child);
+
+            grandChild.markPathToRootAsCheckOnce();
+
+            expect(root.mode).toEqual(CHECK_ALWAYS);
+            expect(disabled.mode).toEqual(DETACHED);
+            expect(parent.mode).toEqual(CHECK_ONCE);
+            expect(child.mode).toEqual(CHECK_ALWAYS);
+            expect(grandChild.mode).toEqual(CHECK_ONCE);
+          });
```
Comment (rkirov) [inline modules/change_detection/test/change_detection_spec.js]: add a test for `markPathToRootAsCheckOnce` handling `CHECK_ONCE` CDs.

File: modules/core/src/compiler/view.js [hunk 7]
```diff
@@ -357,6 +359,8 @@ export class ProtoView {
         lightDom = binder.componentDirective.shadowDomStrategy.constructLightDom(view, childView, element);
         binder.componentDirective.shadowDomStrategy.attachTemplate(element, childView);
 
+        bindingPropagationConfig = new BindingPropagationConfig(view.changeDetector);
```
Comment (rkirov) [inline modules/core/src/compiler/view.js]: why not lazily building it per request, most apps will not use it and it will just create memory pressure on all views.

**Comment by rkirov** [comment]
LGTM, but fix the `markPathToRootAsCheckOnce` handling `CHECK_ONCE` before merging

File: modules/core/src/compiler/view.js (hunk 7 above)
Comment (vsavkin) [inline modules/core/src/compiler/view.js]: I actually started doing it lazily but then decided not to do it. I think it should be done in a separate PR and with measurements (before and after).

File: modules/change_detection/src/abstract_change_detector.js (hunk 5 above)
Comment (vsavkin) [inline modules/change_detection/src/abstract_change_detector.js]: Fixed.

**Comment by vsavkin** [comment]
@mhevery Binder is already very overloaded. I thought maybe BindingPropagator. I am not convinced, however, that we should rename it.

**Comment by vsavkin** [comment]
Merged via https://github.com/angular/angular/commit/c53dce48a902412f7f51759c76fd6b0465285bf6
